### PR TITLE
Prevent eager realization for `dependsOn tasks.withType(...)`

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
Fixes https://github.com/gradle/gradle-native/issues/745
Fixes https://github.com/gradle/gradle/issues/5827

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Flazy%2Fissue-745)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java [hunk 1]
```diff
@@ -156,9 +156,9 @@ public boolean remove(Object o) {
     @Override
     public int size() {
         int i = 0;
-        // TODO this will realize all pending elements
-        for (T o : collection) {
-            if (accept(o)) {
+        // TODO this will realize all pending elements that may be filter out by the filter#spec, still better than realizing everything
```
Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java]: I don't think there is anything we can do about `collection.matching { }` filters as the spec requires a realized element, so this fix is probably as good as we're going to get.

Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java]: We should probably use `iterator()` in `isEmpty()` above as well.

File: subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java [hunk 2]
```diff
@@ -75,8 +75,8 @@ public boolean isEmpty() {
         if (collection.isEmpty()) {
             return true;
         } else {
-            for (T o : collection) {
-                if (accept(o)) {
+            for (Iterator<?> it = iterator(); it.hasNext();) {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java]: Can be replaced by `for(o : this) {}`

Comment (lacasseio) [reply subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java]: Oh, good point. I will change this.

Comment (lacasseio) [reply subprojects/core/src/main/java/org/gradle/api/internal/collections/FilteredCollection.java]: Change done directly on master: https://github.com/gradle/gradle/commit/9f011c89468b96562c329c3a19dc44aa9935330a

**Pull request merged**